server_test.go

```diff
@@ -344,6 +344,105 @@ func TestEndToEnd(t *testing.T) {
 		}, 2*time.Second, 100*time.Millisecond)
 	})
 
+	t.Run("delta IRV support", func(t *testing.T) {
+		const (
+			foo = "foo"
+			bar = "bar"
+			qux = "qux"
+		)
+
+		req := &ads.DeltaDiscoveryRequest{
+			Node:    locator.node,
+			TypeUrl: testResource.TypeURL(),
+		}
+
+		res := new(ads.DeltaDiscoveryResponse)
+
+		// Verify that using InitialResourceVersions prevents the server from resending unchanged resources upon re-subscription.
+		setCacheEntry(t, foo, "0")
+		setCacheEntry(t, bar, "1")
+		req.InitialResourceVersions = map[string]string{
+			foo: "0",
+			bar: "0",
+		}
+		req.ResourceNamesSubscribe = []string{"foo", "bar"}
+		stream, cancel := newStream(t)
+		require.NoError(t, stream.Send(req))
+		waitForResponse(t, res, stream, 10*time.Millisecond)
+		require.Len(t, res.Resources, 1)
+		require.Equal(t, res.Resources[0].Name, "bar")
+
+		// Verify that IRV is being reset.
+		// if IRV is not reset then handler would maintain {"foo": 0, "bar": 0} from previous request on server side
+		// and when empty InitialResourceVersions being pass it would only respond with bar only .
+		// in this case it's responding with foo & bar when the request is being sent on existing stream.
+		req.InitialResourceVersions = map[string]string{}
+		req.ResourceNamesSubscribe = []string{"foo", "bar"}
+		require.NoError(t, stream.Send(req))
+		waitForResponse(t, res, stream, 10*time.Millisecond)
+		require.Len(t, res.Resources, 2)
+		require.ElementsMatch(t, []string{"foo", "bar"}, []string{res.Resources[0].Name, res.Resources[1].Name})
+
+		// Close the stream while the cache is updated to mimic a reconnect
+		cancel()
+
+		// validating for wildcard subscriptions
+		setCacheEntry(t, foo, "0")
+		setCacheEntry(t, qux, "0")
+		bytesCache.Clear(bar, time.Now())
+		req.InitialResourceVersions = map[string]string{
+			foo: "0",
+			bar: "1",
+		}
+		req.ResourceNamesSubscribe = []string{ads.WildcardSubscription}
+		stream, cancel = newStream(t)
+		require.NoError(t, stream.Send(req))
+		waitForResponse(t, res, stream, 10*time.Millisecond)
+		require.Len(t, res.RemovedResources, 1)
+		require.Equal(t, res.RemovedResources[0], bar)
+		require.Len(t, res.Resources, 1)
+		require.Equal(t, res.Resources[0].Name, qux)
+
+		cancel()
+		// reconnect scenario
+		req.InitialResourceVersions = map[string]string{
+			foo: "0",
+			qux: "0",
+		}
```

Comment:
could we also send another request here with an empty IRV, to verify that previous IRV is cleared on server-side , and foo and qux are in the res?

Reply:
oh, i missed it, thanks